resources/assets/coffee/react/beatmapset-page/main.coffee

```diff
@@ -0,0 +1,119 @@
+###
+# Copyright 2016 ppy Pty. Ltd.
+#
+# This file is part of osu!web. osu!web is distributed with the hope of
+# attracting more community contributions to the core ecosystem of osu!.
+#
+# osu!web is free software: you can redistribute it and/or modify
+# it under the terms of the Affero GNU General Public License version 3
+# as published by the Free Software Foundation.
+#
+# osu!web is distributed WITHOUT ANY WARRANTY; without even the implied
+# warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+###
+{div} = React.DOM
+el = React.createElement
+
+BeatmapsetPage.Main = React.createClass
+  mixins: [ScrollingPageMixin]
+
+  getInitialState: ->
+    optionsHash = BeatmapsetPageHash.parse location.hash
+    @initialPage = optionsHash.page
+
+    currentMode = if optionsHash.mode then parseInt optionsHash.mode, 10 else @props.displayedBeatmap
+
+    currentMode: currentMode
+    currentPlaymode: @props.beatmaps[currentMode].mode
```

Comment:
missing check to ensure `currentMode` is valid.

(and why is it called `currentMode` when it actually is `currentBeatmapId`?)